File: integrations/kotlin/src/jsMain/kotlin/angular/ts/Component.kt
package angular.ts

public data class Component<TController> public constructor(
    public val template: String? = null,
    public val templateUrl: String? = null,
    public val controller: InjectableFactory<TController>? = null,
    public val controllerAs: String? = null,
    public val bindings: Map<String, String> = emptyMap(),
    public val transclude: Boolean = false,
    public val require: Map<String, String> = emptyMap(),
)

internal fun Component<*>.toJs(): dynamic {
    val raw = js("{}")

    if (template != null) raw.template = template
    if (templateUrl != null) raw.templateUrl = templateUrl
    if (controller != null) raw.controller = controller.toJs()
    if (controllerAs != null) raw.controllerAs = controllerAs
    if (bindings.isNotEmpty()) raw.bindings = bindings.toJsRecord()
    if (transclude) raw.transclude = true
    if (require.isNotEmpty()) raw.require = require.toJsRecord()

    return raw
}

File: integrations/scala/src/main/scala/angular/ts/Component.scala
package angular.ts

import scala.scalajs.js

final case class Component[A](
    template: js.UndefOr[String] = js.undefined,
    controller: js.UndefOr[InjectableFactory[A]] = js.undefined,
    controllerAs: js.UndefOr[String] = js.undefined,
    templateUrl: js.UndefOr[String] = js.undefined,
    bindings: Map[String, String] = Map.empty,
    replace: js.UndefOr[Boolean] = js.undefined,
    transclude: js.UndefOr[Boolean] = js.undefined,
    require: Map[String, String] = Map.empty,
):
  private[ts] def toJS: js.Object =
    JsObjectBuilder(
      "template" -> template,
      "controller" -> controller.map(_.annotated),
      "controllerAs" -> controllerAs,
      "templateUrl" -> templateUrl,
      "bindings" -> JsObjectBuilder.stringMap(bindings),
      "replace" -> replace,
      "transclude" -> transclude,
      "require" -> JsObjectBuilder.stringMap(require),
    )
